### ISSUE TITLE
accessibility review of Pagination design

### ISSUE THREAD

**Issue by jgiardino** [patternfly/patternfly-next#627]
As part of this issue, review the latest designs for Pagination and provide recommendations about any potential accessibility issues that should be considered in the design or addressed in the html/css implementation.

**Comment by jgiardino** [comment]
We did a group review of the latest design, included here (which is subject to change as soon as I post this comment):
![image](https://user-images.githubusercontent.com/21063328/44540906-5a1f1a00-a6d6-11e8-96b4-8b4f0f8b5569.png)

Additionally, as a separate action item in translating text strings for a modified version of the PF3 pagination, I had posted the runtime html to this [codepen](https://codepen.io/jenngiardino/pen/gdpwRv), and tested in JAWS. I included the text that's announced in JAWS below the pagination controls. We also reviewed this implementation during the meeting.

In looking at the example above, and based on the discussion we had, this is what I think the appropriate read out should/could be. Please note: each bullet represents a separate Down arrow key in JAWS, I'm guessing at some of the JAWS announcements/behaviors based on previous tests, and JAWS announcements of semantics are in [] and sr-only or aria-label text are in ().

1. (Pagination) [navigation]
2. (Items) one (to) ten of ninety-seven (items)
3. (Page Size) [button menu]
4. (Show Previous Page [button unavailable]
5. (Current Page) 1 [edit]
6. (Show Next Page) [button]

Comments during review of the PF4 design centered around the menu for changing page size. In the read-out above, I separated these controls, based on my own personal opinion on what seems to sound best when converting these controls purely to text. The alternative would instead be something like this:

2. (Items) one (to) ten 
3. (Page Size) [button menu]
4. of ninety-seven (items)

The closest component we have currently for this Page Size menu is a dropdown, but the text “1-10” would not make sense to use as the toggle for this menu. Ideally, screen reader users should have a label in the form of aria-label on just the arrow icon that identifies what actions are available in the menu (e.g. “Page Size [button menu]”). In doing so, it might be weird to show the down arrow as the element that has focus (however the current browser handles showing focus) combined with the visual styling to underline both the arrow and the current visible items (i.e. "1 - 10"). 

During review, we also wondered if "Pagination" is a term that users are familiar with.

@mcarrano @kybaker Let me know if you have any questions about these comments so far. These seem to be the things that affect design decisions more so than the remaining comments below.


Some specific comments about html/css implementation:
* Our pf3 implementation uses `<ul>` elements to wrap the buttons, but there seems to be no obvious benefit to doing this, and instead requires more keypresses for the sr user to get through the elements in this component.
* We should wrap the control with `<nav aria-label="Pagination">` so that users understand that these controls are all related and can easily find this control on the page via landmarks.
* Should we use links or buttons? If the page is available via url, then links. If not, then buttons. Our consumers might have both variations, so our implementation might need to support both. 
* It's worth investigating if the page size menu should be implemented using our current dropdown component, or if a custom `select` component should be used. If we use roles `select` and `option`, we can additionally indicate which option is currently selected with `aria-selected`. This attribute expects certain roles, which don't include `menuitem`. (i.e. in the dropdown menu there is no way to indicate an item as selected)
* It's worth investigating further the use of `aria-controls` for this control, which seems applicable here, and determine if it is a better fit for the parent `<nav>` element or individual controls within 


Here's a link that Jan found: http://www.a11ymatters.com/pattern/pagination/ 
The interaction pattern isn't an exact match, but there's some helpful information.

**Comment by mcarrano** [comment]
I understand your concerns with this @jgiardino . This is actually the first time I am seeing this design and my opinion is that combining the 'items per page' selection with reporting the items in view is a bit confusing. In [PatternFly 3](https://www.patternfly.org/pattern-library/navigation/pagination/#design), these are separate controls and I would prefer to keep them that way. I think that allowing the user to set the number of items per page should be an optional feature as that may not be appropriate in all cases. If we adopt the PatternFly 3 approach, would this address the concern or do you think that's still a problem?

As for the question about whether there is a better term to use than "pagination," I'm not sure. Agree that all users wouldn't know what that meant, but can't think of something better.

**Comment by jgiardino** [comment]
> In PatternFly 3, these are separate controls and I would prefer to keep them that way. ... If we adopt the PatternFly 3 approach, would this address the concern or do you think that's still a problem?

I do think isolating the controls will make it easier for screen reader users to understand.

Just FYI, the **Select Example** that we have for the [pf-next form controls](https://pf-next.com/components/FormControls/examples/) seems to be the closest match (in terms of semantics and interaction) to what's shown in the PF3 example. I think if we wanted to use something like the pf-next dropdown, we'd need to make some updates to support the change in behavior (e.g. indicating which option is selected).

BTW, I made some slight changes to the [codepen example](https://codepen.io/jenngiardino/pen/gdpwRv?editors=1000) I referenced above and the text that gets announced. 

As part of these changes, I updated my example to not swap "to" for "-" for screen readers, since it was really kludgy to pull that off anyway, and wasn't outputting complete strings for translation. In testing "-" I noticed that having spaces around the "-" is important for announcing the numbers correctly. For example:

- "11-20" is announced as "one one dash two zero"
- "11 - 20" is announced as "eleven dash twenty"

It might also be worth considering "11 to 20". Although I assume sr users will understand what "dash" means in this context.

**Comment by jgiardino** [comment]
Related to my last statement above:

> It might also be worth considering "11 to 20". Although I assume sr users will understand what "dash" means in this context.

I just saw this in the [Accessibility Documentation](https://docs.google.com/document/d/1q1lBdJaNyTN6BPHOym9zvQ9FJFt-IdXggi75FZZEQPc/edit#heading=h.lli6oty5u2ke) used by the redhat.com team:

> Don't use dashes if you can avoid it. Instead of writing 5–7, write 5 to 7.

**Comment by jgiardino** [comment]
Marking this as closed. 

Also, created an issue for a [custom Select component](https://github.com/patternfly/patternfly-next/issues/708) which will be needed for this design.
